### ISSUE TITLE
learning rate warmup

### ISSUE THREAD

**Issue by chuong98** [williamFalcon/pytorch-lightning#328]
What is the most appropriate way to add learning rate warmup ?
I am thinking about using the hooks.``` def on_batch_end(self):```, but not sure where to put this function to ? Thank you.

**Comment by williamFalcon** [comment]
You can use a [learning rate scheduler](https://pytorch.org/docs/stable/optim.html#torch.optim.lr_scheduler.CosineAnnealingLR) and return it in the [choose_optimizers](https://pytorch-lightning.readthedocs.io/en/latest/LightningModule/RequiredTrainerInterface/#configure_optimizers).

**Comment by magic282** [comment]
Same question here. In Transformer, the LR is adjusted by training step, not epoch. Is there a solution?

**Comment by williamFalcon** [comment]
You can also override ```optimizer_step``` and do it there. Here's an example where the first 500 batches are for warm up.

```python
    def optimizer_step(self, epoch_nb, batch_nb, optimizer, optimizer_i, opt_closure):
        if self.trainer.global_step < 500:
            lr_scale = min(1., float(self.trainer.global_step + 1) / 500.)
            for pg in optimizer.param_groups:
                pg['lr'] = lr_scale * self.hparams.learning_rate

        optimizer.step()
        optimizer.zero_grad()
```

**Comment by Red-Eyed** [comment]
So, I ended up with something like this:
```python
    def configure_optimizers(self):
        optimizer = torch.optim.Adam(self.parameters(), lr=self.hparams.lr)

        def lr_foo(epoch):
            if epoch < self.hparams.warm_up_step:
                # warm up lr
                lr_scale = 0.1 ** (self.hparams.warm_up_step - epoch)
            else:
                lr_scale = 0.95 ** epoch

            return lr_scale

        scheduler = LambdaLR(
            optimizer,
            lr_lambda=lr_foo
        )

        return [optimizer], [scheduler]
```
![image](https://user-images.githubusercontent.com/11485451/84660463-8cd87900-af21-11ea-96d1-d729aad31ac1.png)

**Comment by johngrabner** [comment]
How does one log lr-Adam in tensorboard?  
Using Red-Eyed warmup code and have:
```
def training_step(self, batch, batch_idx):
......
        tensorboard_logs = {'train_loss': loss}
        return {'loss': loss, 'log': tensorboard_logs}
```

**Comment by jdepoix** [comment]
I just stumbled upon this issue, as I was also looking for a way to make my LR scheduler update on each step instead of each epoch. After doing some additional research I found that there is a better way of doing this than overwriting `optimizer_step`. I am guessing this features wasn't available yet when this issue initially cam up, but in version 1.0.3 (don't know the exact version this has been added though) you can just do this:
```python
def configure_optimizers(self):
    optimizer = AdamW(self.parameters(), lr=self.learning_rate)
    scheduler = InverseSquareRootLR(optimizer, self.lr_warmup_steps)
    return [optimizer], [
        {
            'scheduler': scheduler,
            'interval': 'step',
            'frequency': 1,
            'reduce_on_plateau': False,
            'monitor': 'val_loss',
        }
    ]
```

**Comment by btseytlin** [comment]
@jdepoix stumbled upon this when looking for how to make OneCycleLR or CosineAnnealingLR work. You saved me, thank you!

**Comment by shamanez** [comment]
@williamFalcon 

I tried your solution, seems like **self.trainer.global_step**  doesn't get updated.

**Comment by itay1itzhak** [comment]
> I just stumbled upon this issue, as I was also looking for a way to make my LR scheduler update on each step instead of each epoch. After doing some additional research I found that there is a better way of doing this than overwriting `optimizer_step`. I am guessing this features wasn't available yet when this issue initially came up, but in version 1.0.3 (don't know the exact version this has been added though) you can just do this:
> 
> ```python
> def configure_optimizers(self):
>     optimizer = AdamW(self.parameters(), lr=self.learning_rate)
>     scheduler = InverseSquareRootLR(optimizer, self.lr_warmup_steps)
>     return (
>         [optimizer],
>         [
>             {
>                 'scheduler': scheduler,
>                 'interval': 'step',
>                 'frequency': 1,
>                 'reduce_on_plateau': False,
>                 'monitor': 'val_loss',
>             }
>         ]
>     )
> ```

Thanks! Where does the 'InverseSquareRootLR' imported from? I couldn't find it outside of fairseq (which is not usfull for non-fairseq models)?

**Comment by jdepoix** [comment]
@itay1itzhak I implemented that myself, you could use any of the schedulers [implemented by PyTorch](https://pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate)

**Comment by jonaskohler** [comment]
> You can also override `optimizer_step` and do it there. Here's an example where the first 500 batches are for warm up.
> 
> ```python
>     def optimizer_step(self, epoch_nb, batch_nb, optimizer, optimizer_i, opt_closure):
>         if self.trainer.global_step < 500:
>             lr_scale = min(1., float(self.trainer.global_step + 1) / 500.)
>             for pg in optimizer.param_groups:
>                 pg['lr'] = lr_scale * self.hparams.learning_rate
> 
>         optimizer.step()
>         optimizer.zero_grad()
> ```

For me that overwrites the def training_step(), so all of my loss computations, logging etc. get lost. What I did instead was to adjust the learning rate directly in training_step() by replacing _optimizer_ with _self.trainer.optimizers[0]_

**Comment by layumi** [comment]
In my case, the learning rate is linear from 0.1 to 1 in 5000 iterations instead of epochs. 
So it still needs to be in the optimizer step.

```
    def optimizer_step(self, epoch, batch_idx, optimizer, optimizer_idx, optimizer_closure, on_tpu=False, using_native_amp=False, using_lbfgs=False):
        # update params
        optimizer.step(closure=optimizer_closure)

        # manually warm up lr without a scheduler
        if self.trainer.global_step < 5000:
            lr_scale = min(1.0, 0.1 + float(self.trainer.global_step) / 5000.0 )
            for pg in optimizer.param_groups:
                pg["lr"] = lr_scale * self.learning_rate
```

By the way, the optimizer.zero_grad() seems to be not needed, since we have `LightningModule.optimizer_zero_grad`.
